File: sdk/python/ai_assistant/streaming/sse_parser.py
"""
Server-Sent Events (SSE) parser.

Converts the raw byte stream from an ``httpx`` streaming response into a
sequence of ``StreamEvent`` objects.  Handles:

* ``data: {json}\\n\\n`` — standard single-line payloads
* Multi-line ``data:`` fields (concatenated with ``\\n``)
* ``[DONE]`` sentinel signalling end-of-stream
* ``event:`` field (mapped to ``event_type``)
* Empty keep-alive lines (silently skipped)
"""

from __future__ import annotations

import json
import time
from collections.abc import AsyncIterator
from typing import TYPE_CHECKING

from ai_assistant.exceptions import StreamError
from ai_assistant.models.events import EventType, StreamEvent

if TYPE_CHECKING:
    import httpx


class SSEParser:
    """Stateless SSE parser that yields ``StreamEvent`` instances."""

    async def parse(self, response: httpx.Response) -> AsyncIterator[StreamEvent]:
        """Iterate over a streaming ``httpx.Response`` and yield events.

        The caller is responsible for closing the response when done.
        """
        buf_event: str | None = None
        buf_data_lines: list[str] = []

        try:
            async for raw_line in response.aiter_lines():
                line = raw_line.rstrip("\r\n")

                # Empty line = event boundary
                if not line:
                    if buf_data_lines:
                        event = self._flush(buf_event, buf_data_lines)
                        if event is not None:
                            yield event
                    buf_event = None
                    buf_data_lines = []
                    continue

                # Field parsing
                if line.startswith("data:"):
                    value = line[5:].lstrip(" ")
                    buf_data_lines.append(value)
                elif line.startswith("event:"):
                    buf_event = line[6:].strip()
                elif line.startswith(":"):
                    # SSE comment / keep-alive — skip
                    continue

            # Flush any trailing event that was not terminated by a blank line
            if buf_data_lines:
                event = self._flush(buf_event, buf_data_lines)
                if event is not None:
                    yield event

        except Exception as exc:
            raise StreamError(
                f"SSE parse failure: {exc}",
                status_code=None,
            ) from exc

    # -- internal ------------------------------------------------------------

    @staticmethod
    def _flush(event_field: str | None, data_lines: list[str]) -> StreamEvent | None:
        """Parse buffered data lines into a ``StreamEvent``, or ``None`` for sentinels."""
        raw = "\n".join(data_lines)

        # [DONE] sentinel
        if raw.strip() == "[DONE]":
            return StreamEvent(event_type=EventType.DONE, data={}, timestamp=time.time())

        # Parse JSON payload
        try:
            payload: dict = json.loads(raw)
        except json.JSONDecodeError:
            # Non-JSON data line — wrap it
            payload = {"raw": raw}

        # Determine event type: explicit ``event:`` field > payload ``event_type`` > "message"
        event_type = (
            event_field
            or payload.pop("event_type", None)
            or payload.pop("event", None)
            or "message"
        )

        timestamp = payload.pop("timestamp", None)
        if timestamp is None:
            timestamp = time.time()

        return StreamEvent(
            event_type=event_type,
            data=payload,
            timestamp=float(timestamp),
        )

File: sdk/java/src/main/java/com/hejaz/ai/SSEParser.java
package com.hejaz.ai;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

/**
 * Server-Sent Events (SSE) parser.
 *
 * <p>Converts a raw {@link InputStream} (from {@code HttpResponse<InputStream>})
 * into a sequence of {@link StreamEvent} objects. Handles:
 *
 * <ul>
 *   <li>{@code data: {json}\n\n} — standard single-line payloads</li>
 *   <li>Multi-line {@code data:} fields (concatenated with {@code \n})</li>
 *   <li>{@code [DONE]} sentinel signalling end-of-stream</li>
 *   <li>{@code event:} field (mapped to {@code eventType})</li>
 *   <li>Empty keep-alive lines (silently skipped)</li>
 *   <li>SSE comments (lines starting with {@code :})</li>
 * </ul>
 *
 * <p>This parser is stateless and thread-safe — create one instance and
 * reuse it across calls.
 */
public class SSEParser {

    private static final TypeReference<Map<String, Object>> MAP_TYPE = new TypeReference<>() {};

    private final ObjectMapper objectMapper;

    public SSEParser() {
        this(new ObjectMapper());
    }

    public SSEParser(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    /**
     * Parse an SSE stream and deliver events to the consumer.
     *
     * <p>Reads lines from the input stream, buffers until a blank line
     * (event boundary), then flushes the accumulated event. The consumer
     * receives each parsed {@link StreamEvent} in order.
     *
     * <p>Blocks until the stream is exhausted or closed.
     *
     * @param stream  Raw input stream from the HTTP response body.
     * @param handler Consumer that receives each parsed event.
     * @throws StreamException if an I/O or JSON parse error occurs.
     */
    public void parse(InputStream stream, Consumer<StreamEvent> handler) {
        try (var reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            String eventField = null;
            List<String> dataLines = new ArrayList<>();

            String line;
            while ((line = reader.readLine()) != null) {
                // Empty line = event boundary
                if (line.isEmpty()) {
                    if (!dataLines.isEmpty()) {
                        StreamEvent event = flush(eventField, dataLines);
                        if (event != null) {
                            handler.accept(event);
                        }
                    }
                    eventField = null;
                    dataLines = new ArrayList<>();
                    continue;
                }

                // Field parsing per SSE spec
                if (line.startsWith("data:")) {
                    String value = line.substring(5);
                    // Trim single leading space per SSE spec
                    if (value.startsWith(" ")) {
                        value = value.substring(1);
                    }
                    dataLines.add(value);
                } else if (line.startsWith("event:")) {
                    eventField = line.substring(6).strip();
                } else if (line.startsWith(":")) {
                    // SSE comment / keep-alive — skip
                } else if (line.startsWith("id:")) {
                    // Event ID — currently unused but valid SSE field
                } else if (line.startsWith("retry:")) {
                    // Reconnection time — currently unused but valid SSE field
                }
            }

            // Flush any trailing event not terminated by a blank line
            if (!dataLines.isEmpty()) {
                StreamEvent event = flush(eventField, dataLines);
                if (event != null) {
                    handler.accept(event);
                }
            }

        } catch (IOException e) {
            throw new StreamException("SSE stream I/O error: " + e.getMessage(), e);
        }
    }

    // -------------------------------------------------------------------------
    // Internal
    // -------------------------------------------------------------------------

    /**
     * Parse buffered data lines into a {@link StreamEvent}, or null for sentinels.
     */
    private StreamEvent flush(String eventField, List<String> dataLines) {
        String raw = String.join("\n", dataLines);

        // [DONE] sentinel
        if ("[DONE]".equals(raw.strip())) {
            return new StreamEvent(EventType.DONE, Map.of());
        }

        // Parse JSON payload (must be mutable for event_type removal below)
        Map<String, Object> payload;
        try {
            payload = objectMapper.readValue(raw, MAP_TYPE);
        } catch (Exception e) {
            // Non-JSON data line — wrap it (mutable map for consistency)
            payload = new HashMap<>();
            payload.put("raw", raw);
        }

        // Determine event type:
        // 1. Explicit event: field from SSE
        // 2. event_type key in JSON payload
        // 3. event key in JSON payload
        // 4. Fallback to "message"
        String eventType = eventField;
        if (eventType == null) {
            Object et = payload.remove("event_type");
            if (et == null) et = payload.remove("event");
            eventType = et != null ? et.toString() : "message";
        }

        return new StreamEvent(eventType, payload);
    }
}
